src/test/java/io/cdap/plugin/gcp/bigquery/action/BigQueryExecuteTest.java

```diff
@@ -0,0 +1,210 @@
+  BigQueryError bigQueryError;
+  @Mock
+  TableResult queryResults;
+  @Mock
+  JobStatistics.QueryStatistics queryStatistics;
+  @Mock
+  ActionContext context;
+  @Mock
+  StageMetrics stageMetrics;
+  @Mock
+  Metrics metrics;
+  QueryJobConfiguration queryJobConfiguration;
+  BigQueryExecute.Config config;
+  JobInfo jobInfo;
+  JobId jobId;
+  BigQueryExecute bq;
+  MockFailureCollector failureCollector;
+  // Mock error message that will be returned by BigQuery when job fails to execute
+  String mockErrorMessageNoRetry = "Job execution failed with error: $error";
+  String errorMessageRetryExhausted = "Failed to execute BigQuery job. Reason: Retries exhausted.";
+
+  @Before
+  public void setUp() throws InterruptedException, NoSuchMethodException {
+    MockitoAnnotations.initMocks(this);
+    failureCollector = new MockFailureCollector();
+    queryJobConfiguration = QueryJobConfiguration.newBuilder("select * from test").build();
+    config = BigQueryExecute.Config.builder()
+            .setLocation("US").setProject("testProject").setRowAsArguments("false")
+            .setInitialRetryDuration(1L).setMaxRetryDuration(5L)
+            .setMaxRetryCount(1).setRetryMultiplier(2.0).build();
+    jobId = JobId.newBuilder().setRandomJob().setLocation(config.getLocation()).build();
+    jobInfo = JobInfo.newBuilder(queryJobConfiguration).setJobId(jobId).build();
+    bq = new BigQueryExecute(config);
+
+    // Mock Job Creation
+    Mockito.when(bigQuery.create((JobInfo) Mockito.any())).thenReturn(queryJob);
+    Mockito.when(queryJob.waitFor()).thenReturn(queryJob);
+    Mockito.when(queryJob.getStatus()).thenReturn(jobStatus);
+    Mockito.when(jobStatus.getError()).thenReturn(bigQueryError);
+    Mockito.when(bigQueryError.getMessage()).thenReturn(mockErrorMessageNoRetry);
+
+    // Mock Successful Query
+    Mockito.when(queryJob.getQueryResults()).thenReturn(queryResults);
+    Mockito.when(queryResults.getTotalRows()).thenReturn(1L);
+    Mockito.when(queryJob.getStatistics()).thenReturn(queryStatistics);
+    Mockito.when(queryStatistics.getTotalBytesProcessed()).thenReturn(1L);
+
+    // Mock context
+    Mockito.when(context.getMetrics()).thenReturn(stageMetrics);
+    Mockito.doNothing().when(stageMetrics).gauge(Mockito.anyString(), Mockito.anyLong());
+    Mockito.when(stageMetrics.child(Mockito.any())).thenReturn(metrics);
+    Mockito.doNothing().when(metrics).countLong(Mockito.anyString(), Mockito.anyLong());
+
+  }
+
+  @Test
+  public void testExecuteQueryWithExponentialBackoffFailsWithNonRetryError() {
+    Mockito.when(bigQueryError.getReason()).thenReturn("accessDenied");
+    Exception exception = Assert.assertThrows(BigQueryJobExecutionException.class, () -> {
+      bq.executeQueryWithExponentialBackoff(bigQuery, queryJobConfiguration, context);
+    });
+    String actualMessage = exception.getMessage();
+    Assert.assertEquals("Failed to execute BigQuery job. Reason: " + mockErrorMessageNoRetry, actualMessage);
```

Comment:
So, basically we say "Failed to execute BigQuery job. Reason: Job execution failed with error: $error" This is what I want to avoid as it's saying same thing twice

Reply:
Sure, i have removed the wrapping of error message.

Reply:
But this way you are loosing the underlying error. You should either: 
 - do visa versa and pop up only the underlying error message "Job execution failed with error: $error"
  - prepend it with some text that adds value, e.g. "Job retries failed after X tries. Last try failure: Job execution failed with error: $error".

Reply:
Yup I am doing the 1st one , by removing wrapping I ment removing the extra messages string that was present before

Reply:
But you are hiding the underlying "$error", so the only thing that we can see is that retries exhausted without much information about the underlying error.

Reply:
As i already have logging when nth retry is made or the policy is out of retry should i just pop the error and throw that ?

something like this
```java
try {
      Failsafe.with(getRetryPolicy()).run(() -> executeQuery(bigQuery, queryConfig, context));
    } catch (FailsafeException e) {
      throw e.getCause();
    } 
``` 
Also i could drop the `catch (RuntimeException e)`

Reply:
Sure, that's option #1 in the list above

Reply:
I would check if e.getCause is not null as generally not every failure is an exception

Reply:
🤔 
```java
if (e.getCause() != null) {
        throw e.getCause();  
} 
  throw e;
```